source/extensions/filters/network/mysql_proxy/mysql_terminal_filter.cc

```diff
@@ -0,0 +1,522 @@
+              expect_sig.size(), sig.size());
+    return MessageHelper::passwordLengthError(sig.size());
+  }
+  if (expect_sig != sig) {
+    ENVOY_LOG(info, "filter: password is not correct");
+    return MessageHelper::authError(
+        name, parent.read_callbacks_->connection().addressProvider().remoteAddress()->asString(),
+        true);
+  }
+  return absl::optional<ErrMessage>();
+}
+
+Network::FilterStatus MySQLTerminalFilter::onNewConnection() {
+  MySQLMonitorFilter::onNewConnection();
+  downstream_event_handler_->seed = NativePassword::generateSeed();
+  auto packet = MessageHelper::encodeGreeting(downstream_event_handler_->seed);
+  packet.setSeq(0);
+  sendLocal(packet);
+  stepLocalSession(1, MySQLSession::State::ChallengeReq);
+  return Network::FilterStatus::StopIteration;
+}
+
+void MySQLTerminalFilter::DownstreamEventHandler::onNewMessage(MySQLSession::State state) {
+  parent.onNewMessage(state);
+}
+
+void MySQLTerminalFilter::DownstreamEventHandler::onServerGreeting(ServerGreeting&) {
+  ENVOY_LOG(error,
+            "mysql filter: downstream impossible callback onMoreClientLoginResponse is called");
+}
+
+void MySQLTerminalFilter::DownstreamEventHandler::onClientLogin(ClientLogin& login) {
+  parent.onClientLogin(login);
+
+  if (login.isSSLRequest()) {
+    ENVOY_LOG(info, "communication failure due to proxy not support ssl upgrade");
+    auto err = MessageHelper::defaultError("envoy mysql proxy is now not support ssl");
+    parent.onFailure(err, login.getSeq() + 1);
+    return;
+  }
+  ENVOY_LOG(debug, "user {} try to login into database {}", login.getUsername(), login.getDb());
+
+  auto auth_method = AuthHelper::authMethod(login.getClientCap(), login.getAuthPluginName());
+  absl::optional<ErrMessage> err;
+  parent.connect_db_ = login.getDb();
+  switch (auth_method) {
+  case AuthMethod::OldPassword:
+    err = checkAuth(login.getUsername(), login.getAuthResp(),
+                    OldPassword::signature(parent.downstream_password_, seed));
+
+    break;
+  case AuthMethod::NativePassword:
+    err = checkAuth(login.getUsername(), login.getAuthResp(),
+                    NativePassword::signature(parent.downstream_password_, seed));
+    break;
+  default:
+    ENVOY_LOG(info, "auth plugin {} is not support", login.getAuthPluginName());
+    auto auth_switch = MessageHelper::encodeAuthSwitch(seed);
+    auth_switch.setSeq(login.getClientCap() + 1);
+    parent.sendLocal(auth_switch);
+    parent.stepLocalSession(login.getClientCap() + 2, MySQLSession::State::AuthSwitchResp);
+    return;
+  }
+  if (err.has_value()) {
+    parent.onFailure(err.value(), login.getSeq() + 1);
+    return;
+  }
+  tryConnectUpstream(login.getSeq() + 1);
+}
+
+void MySQLTerminalFilter::DownstreamEventHandler::onClientLoginResponse(ClientLoginResponse&) {
+  ENVOY_LOG(error, "mysql filter: onMoreClientLoginResponse impossible callback is called");
```

Comment:
`mysql: onMoreClientLoginResponse callback is not supported.`

Reply:
I am sorry for this message might make you confused, what I want to notice is *in the current implementation, this callback must not be called, because the client login response message is sent by proxy, the downstream decoder will not see this packet*.